activesupport/lib/active_support/cache/key_master.rb

```diff
@@ -0,0 +1,165 @@
+    #   k2.cache_key == k3.cache_key
+    #   # => true
+    #
+    # It is faster to directly add objects to the
+    # KeyMaster object via `<<` than to generate
+    # an intermediate array of objects
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   KeyMaster.new("there").cache_key
+    #   # => "there"
+    #
+    #   KeyMaster.new(["is", "no"]).cache_key
+    #   # => "is/no"
+    #
+    #   KeyMaster.new({ dana: 1, only: 2 }).cache_key
+    #   # => "dana=1/only=2"
+    #
+    #   klass = Class.new { def cache_key; "zuul"; end }
+    #   KeyMaster.new(klass.new).cache_key
+    #   # => "zuul"
+    #
+    # When generating a a cache key, `nil` will be filtered
+    # out, so `["ghost", nil "busters"]` will generate the
+    # same output as ["ghost", "busters"]. If presence of a
+    # `nil` is important, then convert it to a string.
+    #
+    # In addition to generating a `cache_key`
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "patty"; end }
+    #   KeyMaster.new(klass.new).cache_version
+    #   # => "patty"
+    #
+    #   klass = Class.new { def cache_version; "egon"; end }
+    #   KeyMaster.new([klass.new]).cache_version
+    #   # => "egon"
+    #
+    #   klass = Class.new { def cache_version; "venkman"; end }
+    #   KeyMaster.new({foo: klass.new}).cache_version
+    #   # => "venkman"
+    class KeyMaster # :nodoc:
+      attr_reader :cache_key, :cache_version
+
+      def self.to_key(key)
+        return key if key.is_a?(KeyMaster)
+
+        self.new(key)
+      end
+
+      def initialize(key = nil)
+        @cache_key = +""
+        @cache_version = +""
+        self << key
+      end
+
+      def <<(key)
+        return if key.nil?
+
+        if key.respond_to?(:cache_key)
+          store_key(key.cache_key.to_s)
+          store_version(key)
+          return
+        end
+
+        case key
+        when Array
+          key.each do |k|
+            self << k
+          end
+        when Hash
+          key = key.to_a
+          store_version(key)
+          key.sort_by! { |k, _| k.to_s }
+          key.collect! { |k, v| "#{k}=#{v}" }
+
+          key.each do |k|
+            store_key(k)
+          end
+        else
+          store_version(key)
+          store_key(key)
+        end
+      end
+
+      private
+        def store_key(key)
+          @cache_key << "/" unless @cache_key.empty?
```

Comment:
Empty string keys are significant, so checking for emptiness will be insufficient.

Reply:
This line is doing the work of `Array#to_param` which does a `join("/")` on the elements. We want a `/` separator between each entry, but we don't want to start the key with a slash if it's the first time an entry is added.

Reply:
Understood. The implementation is not equivalent, though. Consider adding a key that is an empty string: `@cache_key` will still be empty and the next key that's added will omit the `/` that should be present.